src/backend/common/RenderPass.cpp

```diff
@@ -0,0 +1,103 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "RenderPass.h"
+
+#include "Buffer.h"
+#include "Device.h"
+#include "Texture.h"
+
+namespace backend {
+
+    // RenderPass
+
+    RenderPassBase::RenderPassBase(RenderPassBuilder* builder)
+        : attachments(std::move(builder->attachments)), subpasses(std::move(builder->subpasses)) {
+    }
+
+    nxt::TextureFormat RenderPassBase::GetAttachmentFormat(uint32_t attachment) const {
+        ASSERT(attachment < attachments.size());
+        return attachments[attachment];
+    }
+
+    uint32_t RenderPassBase::GetSubpassCount() const {
+        return subpasses.size();
+    }
+
+    const RenderPassBase::SubpassInfo& RenderPassBase::GetSubpassInfo(uint32_t subpass) const {
+        ASSERT(subpass < subpasses.size());
+        return subpasses[subpass];
+    }
+
+    // RenderPassBuilder
+
+    enum RenderPassSetProperties {
+        RENDERPASS_PROPERTY_SUBPASS_COUNT = 0x1,
+    };
+
+    RenderPassBuilder::RenderPassBuilder(DeviceBase* device)
+        : device(device), subpasses(1) {
+    }
+
+    bool RenderPassBuilder::WasConsumed() const {
+        return consumed;
+    }
+
+    RenderPassBase* RenderPassBuilder::GetResult() {
+        consumed = true;
+        return device->CreateRenderPass(this);
+    }
+
+    void RenderPassBuilder::EnableAttachment(uint32_t attachmentSlot, nxt::TextureFormat format) {
+        if (attachmentSlot > kMaxRenderPassAttachments) {
+            device->HandleError("Render pass attachment index out of bounds");
+            return;
+        }
+        if (attachmentsEnabled[attachmentSlot]) {
+            device->HandleError("Render pass attachment already enabled");
+            return;
+        }
+
+        attachmentsEnabled.set(attachmentSlot);
+        attachments[attachmentSlot] = format;
+    }
+
+    void RenderPassBuilder::SetSubpassCount(uint32_t subpassCount) {
+        if ((propertiesSet & RENDERPASS_PROPERTY_SUBPASS_COUNT) != 0) {
+            device->HandleError("Render pass subpass count property set multiple times");
+            return;
+        }
+
+        subpasses.resize(subpassCount);
+        propertiesSet |= RENDERPASS_PROPERTY_SUBPASS_COUNT;
+    }
+
+    void RenderPassBuilder::SetSubpassColorAttachment(uint32_t subpassIndex, uint32_t outputAttachmentLocation, uint32_t attachmentSlot) {
+        if (subpassIndex >= subpasses.size()) {
+            device->HandleError("Subpass index out of bounds");
+            return;
+        }
+        if (outputAttachmentLocation >= kMaxColorAttachments) {
+            device->HandleError("Output attachment location out of bounds");
+            return;
+        }
+        if (!attachmentsEnabled[attachmentSlot]) {
```

Comment:
I think we can force attachments to be a full and contiguous range so this check disappears.